finagle-stats-core/src/main/scala/com/twitter/finagle/stats/NonReentrantReadWriteLock.scala

```diff
@@ -23,10 +35,41 @@ private[stats] final class NonReentrantReadWriteLock extends AbstractQueuedSynch
   // we haven't provided an implementation for isHeldExclusively because the
   // docs for AQS advise implementors to provide implementations "as applicable"
   // and our use case doesn't require knowing when the lock is held exclusively.
+  val SpinMax: Int = if (rwLockSpinWait() == -1) {
+    // Default number of spins for aarch64 machines set to 5 by default. 
+    // See below comment on reason.
+    if (System.getProperty("os.arch") == "aarch_64") {
+      5
+    } else {
+      1
+    }
+  } else {
+    rwLockSpinWait()
+  }
 
   override def tryAcquireShared(num: Int): Int = {
-    val cur = getState()
-    if (cur >= 0 && compareAndSetState(cur, cur + 1)) 1 else -1
+    var ret: Int = -1
+    var cur: Int = 0
+    var tries: Int = 0
+
+    // Try to acquire the lock up to SpinMax times before failing and 
+    // allowing AbstractQueuedSynchronizer to execute the slow-path lock.
+    // This is important on Arm systems where compareAndSetState may fail
+    // more often due to its weaker memory model.
+    do {
+      cur = getState()
+      if (cur >= 0 && compareAndSetState(cur, cur + 1)) {
+        ret = 1
+      } else {
+        // XXX: When JDK8 support is deprecated, onSpinWait() should be
+        //      used to avoid flooding the memory system with compareAndSetState in
+        //      a tight loop.
+        // Thread.onSpinWait()
+        ret = -1
```

Comment:
Is this assignment necessary? If you want you could put the `tries` increment in this `else` branch.

Reply:
Good point.  Done.